File: arduinoC/libraries/DFRobot_UVIndex240370Sensor/python/DFRobot_UVIndex240370Sensor.py
# -*- coding: utf-8 -*
'''!
  @file       DFRobot_UVIndex240370Sensor.py
  @brief       This is basic library of UVIndex240370Sensor sensor
  @copyright   Copyright (c) 2021 DFRobot Co.Ltd (http://www.dfrobot.com)
  @license     The MIT License (MIT)
  @author      fary(feng.yang@dfrobot.com)
  @version     V1.0
  @date        2024-10-11
  @url         https://github.com/DFRobor/DFRobot_UVIndex240370Sensor
'''

import serial
import time
import smbus
import os
import math
import RPi.GPIO as GPIO
import math

import modbus_tk
import modbus_tk.defines as cst
from modbus_tk import modbus_rtu

class DFRobot_UVIndex240370Sensor():
  I2C_MODE                  = 0x01
  UART_MODE                 = 0x02
  UVINDEX240370SENSOR_DEVICE_ADDR                            =0X23
  UVINDEX240370SENSOR_INPUTREG_UVS_DATA                      =0x06
  UVINDEX240370SENSOR_INPUTREG_UVS_INDEX                     =0x07
  UVINDEX240370SENSOR_INPUTREG_RISK_LEVEL                    =0x08
  UVINDEX240370SENSOR_DEVICE_PID                             =0x427c
  def __init__(self ,bus = 0 ,baud = 9600, mode = I2C_MODE):
    self.mode = 0
    self.resolution = 0
    self.gain = 0
    if mode == self.I2C_MODE:
      self._uart_i2c = self.I2C_MODE
      self.i2cbus = smbus.SMBus(bus)
    else:
      self._uart_i2c = self.UART_MODE
      self.master = modbus_rtu.RtuMaster(serial.Serial(port="/dev/ttyAMA0",baudrate=baud, bytesize=8, parity='N', stopbits=1))
      self.master.set_timeout(1.0)
      
  def begin(self):
    '''!
      @brief Init UVIndex240370Sensor device
      @return Return value init status
    '''
    ret=False
    if self._uart_i2c == self.I2C_MODE:
      buffer = self._read_reg(0x00,2)
      data = buffer[0]|buffer[1]<<8
    else:
      buffer = self._read_reg(0x00,1)
      data = buffer[0]
    if data == self.UVINDEX240370SENSOR_DEVICE_PID:
      ret =True
    return ret

  def read_UV_original_data(self):
    '''!
      @brief Read the UV voltage value
      @return voltage value (mV)
    '''
    if self._uart_i2c == self.I2C_MODE:
      buffer = self._read_reg(self.UVINDEX240370SENSOR_INPUTREG_UVS_DATA,2)
      data = buffer[0]|buffer[1]<<8
    else:
      buffer = self._read_reg(self.UVINDEX240370SENSOR_INPUTREG_UVS_DATA,1)
      data = buffer[0]
    return data

  def read_UV_index_data(self):
    '''!
      @brief Read the UV Index
      @return UV Index
    '''
    if self._uart_i2c == self.I2C_MODE:
      buffer = self._read_reg(self.UVINDEX240370SENSOR_INPUTREG_UVS_INDEX,2)
      data = buffer[0]|buffer[1]<<8
    else:
      buffer = self._read_reg(self.UVINDEX240370SENSOR_INPUTREG_UVS_INDEX,1)
      data = buffer[0]
    return data  


  def read_risk_level_data(self):
    '''!
      @brief Read the risk level
      @return 0-4 (Low Risk,Moderate Risk,High Risk,Very High Risk,Extreme Risk)
    '''
    if self._uart_i2c == self.I2C_MODE:
      buffer = self._read_reg(self.UVINDEX240370SENSOR_INPUTREG_RISK_LEVEL,2)
      data = buffer[0]|buffer[1]<<8
    else:
      buffer = self._read_reg(self.UVINDEX240370SENSOR_INPUTREG_RISK_LEVEL,1)
      data = buffer[0]
    return data

class DFRobot_UVIndex240370Sensor_I2C(DFRobot_UVIndex240370Sensor):
  '''!
    @brief An example of an i2c interface module
  '''
  def __init__(self ,bus):
    self._addr = self.UVINDEX240370SENSOR_DEVICE_ADDR
    super().__init__(bus,0,self.I2C_MODE)   
    
  
  def _read_reg(self, reg_addr ,length):
    '''!
      @brief read the data from the register
      @param reg_addr register address
      @param length read data
    '''
    self._reg = reg_addr
    rslt = self.i2cbus.read_i2c_block_data(self._addr ,self._reg , length)
    return rslt
       

class DFRobot_UVIndex240370Sensor_UART(DFRobot_UVIndex240370Sensor):
  '''!
    @brief An example of an UART interface module
  '''
  def __init__(self):
    self._baud = 9600
    self._addr = self.UVINDEX240370SENSOR_DEVICE_ADDR
    try:
      super().__init__(0,self._baud,self.UART_MODE)
    except:
      print ("plese get root!")
   
  
  def _read_reg(self, reg_addr ,length):
    '''!
      @brief Read data from the sensor
    '''
    return list(self.master.execute(self._addr, cst.READ_INPUT_REGISTERS, reg_addr, length))
  

File: arduinoC/libraries/DFRobot_UVIndex240370Sensor/DFRobot_UVIndex240370Sensor.cpp
/*!
 * @file DFRobot_UVIndex240370Sensor.cpp
 * @brief This is the method implementation file of UVIndex240370Sensor
 * @copyright   Copyright (c) 2010 DFRobot Co.Ltd (http://www.dfrobot.com)
 * @license     The MIT License (MIT)
 * @author [fary](feng.yang@dfrobot.com)
 * @version  V0.1.0
 * @date  2025-05-14
 * @url https://github.com/lekeopen/uvindex240370sensor
 */

#include "DFRobot_UVIndex240370Sensor.h"

DFRobot_UVIndex240370Sensor::DFRobot_UVIndex240370Sensor(TwoWire *pWire)
{
  _pWire = pWire;
};

DFRobot_UVIndex240370Sensor::DFRobot_UVIndex240370Sensor(Stream *s) : DFRobot_RTU(s)
{
  _s = s;
}

bool DFRobot_UVIndex240370Sensor::begin(void)
{
  setTimeoutTimeMs(200);
  bool ret = false;
  if (_pWire)
  {
    _pWire->begin();
    _pWire->beginTransmission(_addr);
    if (_pWire->endTransmission() == 0)
    {
      ret = true;
    }
  }
  else
  {
    uint8_t buffer[2];
    readReg(0x00, buffer, 2);
    uint16_t data = (uint16_t)buffer[0] << 8 | buffer[1];
    if (data == UVINDEX240370SENSOR_DEVICE_PID)
    {
      ret = true;
    }
  }
  return ret;
}

uint16_t DFRobot_UVIndex240370Sensor::readUvOriginalData(void)
{
  uint8_t buffer[2];
  readReg(UVINDEX240370SENSOR_INPUTREG_UVS_DATA, buffer, 2);
  uint16_t data = (uint16_t)buffer[0] << 8 | buffer[1];
  return data;
}

uint16_t DFRobot_UVIndex240370Sensor::readUvIndexData(void)
{
  // 获取原始UV值
  uint16_t rawValue = readUvOriginalData();

  // 对原始值进行映射计算UV指数
  // 小于100的值映射为0
  // 参考UV指数标准: 0-2低风险，3-5中等风险，6-7高风险，8-10极高风险，11+极端风险
  uint16_t uvIndex;

  if (rawValue < 50)
  {              // 将阈值从100调整为50，与Python版本保持一致
    uvIndex = 0; // 非常小的值视为UV指数0
  }
  else if (rawValue < 227)
  {              // 将阈值从200调整为227，与Python版本保持一致
    uvIndex = 1; // 较小的值视为UV指数1
  }
  else if (rawValue < 318)
  {              // 将阈值从400调整为318，与Python版本保持一致
    uvIndex = 2; // 适中的值视为UV指数2
  }
  else if (rawValue < 408)
  {
    uvIndex = 3; // UV指数3
  }
  else if (rawValue < 503)
  {
    uvIndex = 4; // UV指数4
  }
  else if (rawValue < 606)
  {
    uvIndex = 5; // UV指数5
  }
  else if (rawValue < 696)
  {
    uvIndex = 6; // UV指数6
  }
  else if (rawValue < 795)
  {
    uvIndex = 7; // UV指数7
  }
  else if (rawValue < 881)
  {
    uvIndex = 8; // UV指数8
  }
  else if (rawValue < 976)
  {
    uvIndex = 9; // UV指数9
  }
  else if (rawValue < 1079)
  {
    uvIndex = 10; // UV指数10
  }
  else
  {
    uvIndex = 11; // 超过1079的值视为UV指数11
  }

  return uvIndex;
}
uint16_t DFRobot_UVIndex240370Sensor::readRiskLevelData(void)
{
  // 获取我们计算后的UV指数
  uint16_t uvIndex = readUvIndexData();

  // 根据UV指数计算风险级别
  // 0-2: 低风险(0)
  // 3-5: 中等风险(1)
  // 6-7: 高风险(2)
  // 8-10: 非常高风险(3)
  // 11+: 极端风险(4)
  uint16_t riskLevel;

  if (uvIndex < 3)
  {
    riskLevel = 0; // 低风险
  }
  else if (uvIndex < 6)
  {
    riskLevel = 1; // 中等风险
  }
  else if (uvIndex < 8)
  {
    riskLevel = 2; // 高风险
  }
  else if (uvIndex < 11)
  {
    riskLevel = 3; // 非常高风险
  }
  else
  {
    riskLevel = 4; // 极端风险
  }

  return riskLevel;
}

uint8_t DFRobot_UVIndex240370Sensor::readReg(uint16_t reg, void *pBuf, uint8_t size)
{
  uint8_t *_pBuf = (uint8_t *)pBuf;
  uint8_t _reg = 0;
  if (pBuf == NULL)
  {
    DBG("data error");
    return 0;
  }
  if (_pWire)
  {
    _pWire->beginTransmission(_addr);
    _pWire->write(reg);
    _pWire->endTransmission();
    _pWire->requestFrom(_addr, size);
    for (uint8_t i = 0; i < size; i++)
      _pBuf[i] = _pWire->read();
    for (uint8_t i = 0; i < size;)
    {
      uint8_t temp = _pBuf[i];
      _pBuf[i] = _pBuf[i + 1];
      _pBuf[i + 1] = temp;
      i += 2;
    }
    return size;
  }
  else
  {
    return readInputRegister(_addr, reg, _pBuf, size);
  }
}
